test/jdk/java/lang/invoke/MethodHandleProxies/WrapperHiddenClassTest.java

```diff
@@ -0,0 +1,212 @@
+                cob.putstatic(CD_HostileWrapper, TYPE, CD_Class);
+                cob.return_();
+            });
+
+            // <init>
+            clb.withMethodBody(INIT_NAME, MTD_void, ACC_PUBLIC, cob -> {
+                cob.aload(0);
+                cob.invokespecial(CD_Object, INIT_NAME, MTD_void);
+                cob.return_();
+            });
+
+            // implementation
+            clb.withMethodBody("compare", MTD_int_Object_Object, ACC_PUBLIC, cob -> {
+                cob.aload(1);
+                cob.checkcast(CD_Integer);
+                cob.aload(2);
+                cob.checkcast(CD_Integer);
+                cob.invokestatic(CD_Integer, "compareTo", MTD_int_Integer);
+                cob.ireturn();
+            });
+        });
+        var l = MethodHandles.lookup().defineHiddenClass(bytes, true);
+        return (Comparator<Integer>) l.findConstructor(l.lookupClass(), MethodType.methodType(void.class)).invoke();
+    }
+
+    /**
+     * Ensures a user interface cannot access a Proxy implementing it.
+     */
+    @Test
+    public void testNoAccess() {
+        var instance = asInterfaceInstance(Client.class, MethodHandles.zero(void.class));
+        var instanceClass = instance.getClass();
+        var interfaceLookup = Client.lookup();
+        assertEquals(MethodHandles.Lookup.ORIGINAL, interfaceLookup.lookupModes() & MethodHandles.Lookup.ORIGINAL,
+                "Missing original flag on interface's lookup");
+        assertThrows(IllegalAccessException.class, () -> MethodHandles.privateLookupIn(instanceClass,
+                interfaceLookup));
+    }
+
+    /**
+     * Tests the Proxy module properties for Proxies implementing system and
+     * user interfaces.
+     */
+    @ParameterizedTest
+    @ValueSource(classes = {Client.class, Runnable.class})
+    public void testModule(Class<?> ifaceClass) {
+        var mh = MethodHandles.zero(void.class);
+
+        var inst = asInterfaceInstance(ifaceClass, mh);
+        Module ifaceModule = ifaceClass.getModule();
+        Class<?> implClass = inst.getClass();
+        Module implModule = implClass.getModule();
+
+        String implPackage = implClass.getPackageName();
+        assertFalse(implModule.isExported(implPackage),
+                "implementation should not be exported");
+        assertTrue(ifaceModule.isExported(ifaceClass.getPackageName(), implModule),
+                "interface package should be exported to implementation");
+        assertTrue(implModule.isOpen(implPackage, MethodHandleProxies.class.getModule()),
+                "implementation class is not reflectively open to MHP class");
+        assertTrue(implModule.isNamed(), "dynamic module must be named");
+        assertTrue(implModule.getName().startsWith("jdk.MHProxy"),
+                () -> "incorrect dynamic module name: " + implModule.getName());
+
+        assertSame(implClass.getClassLoader(), implModule.getClassLoader(),
```

Comment:
I think this should check against `ifaceModule.getClassLoader()` instead, right? Since the dynamic module is defined in the interface' class loader.

Reply:
Good catch!   Yes, this should check against the interface's loader.

Reply:
I've added a new check for `ifaceClass.getClassLoader()`, should be functionally equivalent. This check is originally part of `ProxyForMethodHandle` test, so I decided to keep it as well as it tests impl module's behavior.